### PR TITLE
graph,graphql,server,store: Subgraph Sql Service

### PR DESCRIPTION
**Pull request by gusinacio**
Semiotic Labs is thrilled to present the Subgraph SQL Service in partnership with TheGuild and Edge & Node. This exposes a GraphQL type to any graph-node user, providing a secure SQL interface to directly query a subgraph's entities.

## SQL Service

We introduce a new sql field on the "Query" type that exposes inputs so graph-node can receive raw SQL.

We also implemented a lot of safe sql rewrite to remove functions that we think may break the postgres instance along with some schema safety.

We also expose each of the tables via Common Table Expressions where we can have granular control over what is exposed or not and some workaround exposing only the latest block data.

## What we cover on this PR

- New `sql` query type
- Defined API for inputs for backward compatibility with future work
- Environment variable/Config to enable or disable the SQL service
- Safety measures for SQL threats via complete AST SQL scan.
- Latest block data
- Single subgraph query
- JSON and CSV `union` output

## What we do not cover on this PR

- Cross subgraph joins
- Query on specific block/snapshot
- Query optimizations
- Query parameters binding
- Query/Join between two different blocks/snapshots

### CODE REVIEW

File: graph/src/schema/api.rs
Comment (suchapalaver) [inline graph/src/schema/api.rs:1347]:
```suggestion
```

File: graph/src/data/graphql/queryable_type.rs
Comment (gusinacio) [reply graph/src/data/graphql/queryable_type.rs:29]: Rust has exhaustive pattern matching, but it all falls apart when we use the catch-all argument.

We use all the possible enum values so in case we add another one, the compiler will tell us where it's being used.

Comment (suchapalaver) [reply graph/src/data/graphql/queryable_type.rs:29]: That makes total sense, thanks!

**Comment by lutter** [comment]
Wow .. that's a really nice addition!

Before I start reviewing this in detail, I noticed a few things from a quick look at the PR:
* there is no end-user documentation. Can you add something to `docs/` that explains how to run queries? Obvious questions that should answer are: how do I send a query (example)? how do I use bind params? how does the GraphQL schema relate to the tables that this can query? What are the attributes/columns that can be queried? How much of SQL does this cover? I haven't looked but does this allow e.g., aggregations, window functions, CTEs etc.?
* for operators: since SQL queries can be arbitrarily complex and take an arbitrary amount of time, how do they guard against long-running queries?
* a minor nit is that commit messages should follow the guidelines [here](https://github.com/graphprotocol/graph-node/blob/master/CONTRIBUTING.md#commit-messages-and-pull-requests) This won't cause us to not approve the PR, but it would be nice to keep with the current scheme.
* this might be too much work, but I see that a `sql` crate appears and is then folded into the `store` - would it be possible to squash various commits so that there are fewer commits that fix up previous ones, not just for that crate but other fix commits, too? It'll make reviewing a lot easier

**Comment by gusinacio** [comment]
> Wow .. that's a really nice addition!
> 
> Before I start reviewing this in detail, I noticed a few things from a quick look at the PR:
> 
> * there is no end-user documentation. Can you add something to `docs/` that explains how to run queries? Obvious questions that should answer are: how do I send a query (example)? how do I use bind params? how does the GraphQL schema relate to the tables that this can query? What are the attributes/columns that can be queried? How much of SQL does this cover? I haven't looked but does this allow e.g., aggregations, window functions, CTEs etc.?
> * for operators: since SQL queries can be arbitrarily complex and take an arbitrary amount of time, how do they guard against long-running queries?
> * a minor nit is that commit messages should follow the guidelines [here](https://github.com/graphprotocol/graph-node/blob/master/CONTRIBUTING.md#commit-messages-and-pull-requests) This won't cause us to not approve the PR, but it would be nice to keep with the current scheme.
> * this might be too much work, but I see that a `sql` crate appears and is then folded into the `store` - would it be possible to squash various commits so that there are fewer commits that fix up previous ones, not just for that crate but other fix commits, too? It'll make reviewing a lot easier

Hey @lutter, thank you for the information. Here are some notes:
- we added end-user documentation about the SQL-service to `/docs` as requests
- graph-node has some build-in mechanisms to cancel long-running queries and we are relying on that
- thanks for pointing out the guidelines, we rebased and reworded all the commits to make it easier to review
- while rebasing, we removed how it was made and created commits of the "happy-path". We removed this appearance and fold of the SQL crate and fix up some commits.

File: docs/sql_service.md
Comment (lutter) [inline docs/sql_service.md:63]: If you set this to `true` to turn it on, it would make sense to say that the default is `false`

File: docs/sql_service.md
Comment (lutter) [inline docs/sql_service.md:114]: This does not explain at all how I go from the GraphQL subgraph schema to the schema against which I can execute queries. If I have an entity type `DailyPositionSnapshot`, what's the table I query? Does it matter if that type is mutable, immutable, or a timeseries?

To avoid locking ourselves into how data is currently stored, I would suggest the following:
* the table for an `@entity` type is the name of the type in snakecase
* the columns of the table are all the non-derived attributes of the type, also snakecased
* for aggregations, the table name is the name of the type snakecased together with the interval, i.e. `type Stats @aggregation(intervals: ["hour", "day"], source: "Data" { .. }` becomes `stats('hour')` in a SQL query
* queries are always executed at a specific block, determined by the block constraint on the `sql` element

File: graph/src/data/graphql/queryable_type.rs
Comment (lutter) [inline graph/src/data/graphql/queryable_type.rs:11]: I like the name `QueryableType` much better than `ObjectOrInterface`; as I understand it, adding `union` here is necessary because of the return type of the `sql` query element, right?

**Review by lutter — CHANGES_REQUESTED** [review]
I am not entirely through reviewing this, but didn't want to hold the review back for even longer.

Besides the comments I made inline, one major missing piece are end-to-end tests. For GraphQL queries we have a pretty comprehensive test suite in `store/test-store/tests/core/interfaces.rs` and `store/test-store/tests/graphql/query.rs`. I would want a similarly comprehensive test suite for SQL queries - basically if a change to `graph-node` breaks SQL queries, I want to find out about that from a failing test similar to the ones for GraphQL.

File: graphql/src/store/resolver.rs
Comment (gusinacio) [reply graphql/src/store/resolver.rs:418]: the reason why we have this check here is in the remote case where the schema contains a type called `Sql` which a disabled SQL service graph-node would allow it.

File: graphql/src/store/resolver.rs
Comment (gusinacio) [reply graphql/src/store/resolver.rs:330]: This is good to know. I got to this crate when I was developing but at the time graph-node wasn't using diesel v2. Gonna take a look at it

File: Cargo.toml
Comment (gusinacio) [reply Cargo.toml:34]: rustfmt did this, gonna revert it.

File: store/postgres/src/sql/constants.rs
Comment (gusinacio) [reply store/postgres/src/sql/constants.rs:249]: We were not relying on the fact of being deterministic. Should this PR also try to achieve this?

File: store/postgres/src/sql/parser.rs
Comment (gusinacio) [reply store/postgres/src/sql/parser.rs:35]: Within internal tests, users struggle to understand that they can not use the `0x` prefix and tend to think that the data isn't there. Also, users believed that the UX was better when they received the prefix because they could ensure that it was an address.

It's easy to rewrite every string that starts with `0x` with `\x`, but if I'm not mistaken graph-node the responses from graphql queries are formatted somehow, we probably need to use it as well.

File: store/postgres/src/sql/parser.rs
Comment (gusinacio) [reply store/postgres/src/sql/parser.rs:52]: One of the requirements is that we didn't want to expose all the data and be able to "filter" what is exposed via SQL or not(i.e $block, etc). We had 3 possible solutions at the time:
- Use AST to keep track of all columns and reject via static analysis.
- Create Views
- Create CTEs

The AST route is by far the most efficient and most difficult to implement because we need to keep track of column names, possible rewrites and aliases, etc, and given the timeframe we decided to let it for future work. We had concerns with the view approach because we would need to handle "migrations" and update it for already deployed subgraphs.

The CTE route was the easiest and simplest way to create a "VIEW" per query and also rewrite/filter what we want.

File: store/postgres/src/sql/parser.rs
Comment (gusinacio) [reply store/postgres/src/sql/parser.rs:76]: Thanks for the heads up. It's in the details where the code familiarity shines.

File: store/postgres/src/sql/parser.rs
Comment (gusinacio) [reply store/postgres/src/sql/parser.rs:44]: Thanks!

File: graph/src/data/graphql/queryable_type.rs
Comment (gusinacio) [reply graph/src/data/graphql/queryable_type.rs:11]: Exactly! we want the user to select which type of response: JSON or CSV. This is also future-proof for possible new formats.

File: docs/sql_service.md
Comment (tumaysem) [reply docs/sql_service.md:114]: Thanks, you (@lutter) have already summed up some part of it. Let's expand on this with examples.
Can you give a subgraph example with aggregation ?

File: store/postgres/src/sql/constants.rs
Comment (lutter) [reply store/postgres/src/sql/constants.rs:249]: The problem is that it is really hard to take things away - unless there is a very strong need for something non-deterministic, I would avoid that for now.

File: store/postgres/src/sql/parser.rs
Comment (lutter) [reply store/postgres/src/sql/parser.rs:35]: You can't rewrite every `0x` string because quite a few subgraphs store hex data as strings and may or may not use a `0x` prefix for that - for strings, we don't know that as it's completely under the control of the mappings.

But this particular rewrite is an absolute performance killer. It will even rewrite expressions like `a.id = b.some_col`; that makes it impossible for Postgres to use the BTree index on those two columns, and might cause a full table scan depending in what other clauses the query contains.

For the UX problem, I see two options:
* document better that hex constants need to be written as `\xdeadbeef`
* do a bit of type inference on the AST; you'd want to infer for an expression like `a.col = '0x..'` that the literal needs to be a `bytea` based on the type of `a.col` and rewrite the literal in those cases. It's a little delicate since if you don't catch all cases (e.g., you also need to cover stuff like `a.col in ('0x..', ..)` and a lot of other expressions) it will confuse users even more since `0x..` will work in some places but not in others

There's another nuance of how indexes and queries interact: for most string and bytea columns, we have an index on a prefix of the value, not the whole value since that can be too big to put into an index. For GraphQL queries, the machinery that figures out whether to use a prefix or the whole column starts [here](https://github.com/graphprotocol/graph-node/blob/master/store/postgres/src/relational_queries.rs#L1464-L1473) If the query is not transformed to take advantage of those indexes, it again might become unbearably slow.

File: store/postgres/src/sql/parser.rs
Comment (lutter) [reply store/postgres/src/sql/parser.rs:52]: Yes, if the CTE is used only once, it gets inlined; if it is used more than once, it is materialized. Materializing these CTEs in general will be very expensive.

What I had in mind was rewriting something like `select * from tokens` to `select * from (select * from tokens where block_range @> $block) as tokens` with the hope that the Postgres optimizer can rewrite the query to move the where clause to the outer query if that seems to improve things, though I haven't tested that.

For the example queries you tried, how common are queries that use the same CTE more than once?

File: store/postgres/src/sql/validation.rs
Comment (lutter) [reply store/postgres/src/sql/validation.rs:49]: This needs to be addressed before this PR can be merged, as it makes it possible to destroy data.

File: docs/sql_service.md
Comment (lutter) [reply docs/sql_service.md:114]: [This subgraph](https://github.com/graphprotocol/common-subgraphs/tree/main/aggregation-example) is kinda silly because it just aggregates block numbers, but it explains how to use aggregations in various ways. The [aggregation docs](https://github.com/graphprotocol/graph-node/blob/master/docs/aggregations.md) also have some more realistic examples.

File: store/postgres/src/sql/parser.rs
Comment (tumaysem) [reply store/postgres/src/sql/parser.rs:52]: Actually heavy as all CTEs are mapping to tables. On the other hand this is exactly conforms the non recursive and side-effect-free. So I expect optimizer to inline the CTE and merge filters. I haven't checked the execution plans but performance wise everything seems to be okey.

**Pull request closed without merging**
